src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs

```diff
@@ -388,7 +389,8 @@ public override BoundStatement InstrumentPatternSwitchStatement(BoundPatternSwit
 
         public override BoundStatement InstrumentPatternSwitchWhenClauseConditionalGotoBody(BoundExpression original, BoundStatement ifConditionGotoBody)
         {
-            WhenClauseSyntax whenClause = (WhenClauseSyntax)original.Syntax.Parent;
+            WhenClauseSyntax whenClause = original.Syntax.FirstAncestorOrSelf<WhenClauseSyntax>();
```

Comment:
>WhenClauseSyntax whenClause = original.Syntax.FirstAncestorOrSelf<WhenClauseSyntax>(); [](start = 12, length = 86)

It feels like something similar should be done with syntax in DynamicAnalysisInjector. Also, consider adding a test for dynamic analysis.